### ISSUE TITLE
Proposal: change plugin structure, don't use property on function

### ISSUE THREAD

**Issue by JounQin** [stylelint/stylelint#7388]
### What is the problem you're trying to solve?

https://github.com/csstools/stylelint-use-logical/pull/22#issuecomment-1849022082

> It doesn't work well with TypeScript by default.

### What solution would you like to see?

> I think we should add back the previous support for compatibility, and export `default` + `ruleName` should be preferred IMO.

**Comment by ybiquitous** [comment]
@JounQin You seem to misunderstand. We have supported the rule function & property form. 16.0.0 does NOT break the compatibility. See below:

- https://github.com/stylelint/stylelint/blob/15.11.0/docs/developer-guide/plugins.md#the-anatomy-of-a-plugin
- https://github.com/stylelint/stylelint/blob/15.11.0/lib/rules/alpha-value-notation/index.js#L224

**Comment by JounQin** [comment]
@ybiquitous 

My proposal is not using such structure but change to use different exports:

```ts
export default () => {}

export const { ruleName }
```

**Comment by ybiquitous** [comment]
> My proposal is not using such structure but change to use different exports:

Your proposal would allow another form, right? If so, I don't think it's worth complicating the codebase. We should support both implementations, tests, docs, etc.

---

> Why `stylelint.createPlugin` call does not set `ruleName` property for `ruleFunction` automatically if it has not been set?

I'm also unsure about *why* (historical reason?). But it has never caused a big problem, and I don't think it will be a big one in the future. Instead, allowing both ways may confuse plugin authors.

**Comment by JounQin** [comment]
> Your proposal would allow another form, right?

Yes, because I think the current form is not a good practice IMO.

>> It doesn't work well with TypeScript by default.

> I'm also unsure about why (historical reason?).

It's already causing problems like the referenced issue https://github.com/csstools/stylelint-use-logical/pull/22#issuecomment-1849017589

The developer should not care about such tiny detail which is a bad practice even further.

**Comment by romainmenke** [comment]
> It's already causing problems like the referenced issue https://github.com/csstools/stylelint-use-logical/pull/22#issuecomment-1849017589

I don't think we must attribute too much weight to this single failure.
The plugin was first created 5 years ago and wasn't very actively maintained.

**Comment by JounQin** [comment]
> I don't think we must attribute too much weight to this single failure.

I'm just pointing a reported issue, there could be other unmigrate plugins.

And my point is not only compatibility, but the practice.

I think we can have two directions:

1. keep only current one form, but support add `ruleName` property automatically in `stylelint.createPlugin`, so that the developer does not to care about it, and won't have any issue when use TypeScript.
2. add previous support for `export default` and `export const ruleName` at the same time.

**Comment by romainmenke** [comment]
> add **previous support** 

I think what everyone is saying is that this was never really supported.
I don't think it was ever part of either documentation, guides, blogs posts, ...

I think it worked by accident/mistake and that someone stumbled over this by mistake.

That is why I do not think we must assign too much weight to a single failure.
That makes it seem as if changing this would be restoring a previous feature.

I don't think that is true.
I think it must be considered as a completely new API addition.

-----------

> `export default` and `export const ruleName` at the same time.

We are having a lot of problems with all the PostCSS plugins over at csstools because we need to define exports like these for PostCSS.

TypeScript has issues with mixed default and named exports.

I think it would be a mistake to adopt an API that promotes this.

**Comment by JounQin** [comment]
> TypeScript has issues with mixed default and named exports.

Although I don't understand what exactly do you mean, TypeScript just works as the ES spec standard, but I think we still can move forward to direction 1?

**Comment by JounQin** [comment]
A new issue is just raised #7405

**Comment by Jordan-Hall** [comment]
Between both plugins they over 30,000 weekly download. I agree its not actively maintain hence the fork took off and I fixed the spec issues. I'm happy to rewrite it but ideally would be good to get 16 support in ASAP then I'll recreate with better standard of code

**Comment by jeddy3** [comment]
@Jordan-Hall I believe you can just fix your `require` statements for 16 support:

```js
ruleFunction.ruleName = ruleName;
ruleFunction.messages = messages;
ruleFunction.meta = meta;

module.exports = stylelint.createPlugin(ruleName, ruleFunction);
```

It'll be a patch release as it'll be backwards compatible with older versions of Stylelint.

You can then migrate to ESM at your leisure, which will drop support for Stylelint less than 16.

**Comment by jeddy3** [comment]
I just saw [you've patched the plugin already](https://github.com/Jordan-Hall/stylelint-use-logical-spec/commit/1c47ae4b235365dd653f37640730b613e877108e). Thank you for updating your plugin.

I believe we can close this issue now. As @ybiquitous [said](https://github.com/stylelint/stylelint/issues/7388#issuecomment-1849151026), let's stick with the form we've always had.

**Comment by JounQin** [comment]
@jeddy3 

Can we make it better?

> keep only current one form, but support add ruleName property automatically in stylelint.createPlugin, so that the developer does not to care about it, and won't have any issue when use TypeScript.
